sdks/java/io/mqtt/src/test/java/org/apache/beam/sdk/io/mqtt/MqttIOTest.java

```diff
@@ -0,0 +1,189 @@
+import org.apache.beam.sdk.values.PCollection;
+import org.eclipse.paho.client.mqttv3.IMqttDeliveryToken;
+import org.eclipse.paho.client.mqttv3.MqttCallback;
+import org.eclipse.paho.client.mqttv3.MqttClient;
+import org.eclipse.paho.client.mqttv3.MqttException;
+import org.eclipse.paho.client.mqttv3.MqttMessage;
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/**
+ * Tests of {@link MqttIO}.
+ */
+@RunWith(JUnit4.class)
+public class MqttIOTest implements Serializable {
+
+  private static transient BrokerService broker;
+
+  @BeforeClass
+  public static void startBroker() throws Exception {
+    broker = new BrokerService();
+    broker.setUseJmx(false);
+    broker.setPersistenceAdapter(new MemoryPersistenceAdapter());
+    broker.addConnector(new URI("mqtt://localhost:11883"));
+    broker.start();
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testRead() throws Exception {
+    Pipeline pipeline = TestPipeline.create();
+
+    PCollection<byte[]> output = pipeline.apply(
+        MqttIO.read()
+            .withMqttConnectionConfiguration(
+                MqttIO.MqttConnectionConfiguration.create(
+                    "tcp://localhost:11883",
+                    "BEAM_PIPELINE",
+                    "READ_TOPIC"))
+          .withMaxNumRecords(10));
+
+    PAssert.thatSingleton(output.apply("Count", Count.<byte[]>globally()))
+        .isEqualTo(10L);
+    PAssert.that(output).satisfies(new SerializableFunction<Iterable<byte[]>, Void>() {
+      @Override
+      public Void apply(Iterable<byte[]> input) {
+        int count = 0;
+        for (byte[] element : input) {
+          String inputString = new String(element);
+          Assert.assertEquals("This is test " + count, inputString);
+          count++;
+        }
+        return null;
+      }
+    });
+
+    // produce messages on the broker in another thread
+    // This thread prevents to block the pipeline waiting for new messages
+    Thread thread = new Thread() {
+      public void run() {
+        try {
+          // gives time to the pipeline to start
+          Thread.sleep(2000);
+        } catch (Exception e) {
+          // nothing to do
+        }
+        try {
+          MqttClient client = new MqttClient("tcp://localhost:11883", "publisher");
+          client.connect();
+          for (int i = 0; i < 10; i++) {
+            MqttMessage message = new MqttMessage();
+            message.setQos(2);
+            message.setRetained(true);
+            message.setPayload(("This is test " + i).getBytes());
+            client.publish("READ_TOPIC", message);
+          }
+          client.disconnect();
+          client.close();
+        } catch (Exception e) {
+          // nothing to do
+        }
+      }
+    };
+    thread.start();
```

Comment:
You need to thread.start() to start the thread, and then thread.join() to join it (i.e. wait for it to complete). Sounds like you called join() without calling start(), which deadlocked because the thread wasn't started.